### PR TITLE
Implemented straitforward closest element search in a sorted list using binary search

### PR DESCRIPTION
**Pull request by gravikovich**

### CODE REVIEW

**Comment by eitanhs** [comment]
I'm not sure I like it more than what I've suggested (lambda for substraction) since the user might think she can pass whatever object type she wants...
Also, why not at least add an IComperable constraint?

**Comment by gravikovich** [comment]
> I'm not sure I like it more than what I've suggested (lambda for substraction) since the user might think she can pass whatever object type she wants...
> Also, why not at least add an IComperable constraint?

1. The lambda is quite tedious for the user, the usage now is very clear and works just as BinarySearch does. Also it's not trivial, for example `DateTime - DateTime = TimeSpan`
1. https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.binarysearch?view=netcore-3.1 - just as binary search is unconstrained and uses the default comparer

File: Bsii.Dotnet.Utils/Collections/CollectionExtensions.cs [hunk 1]
```diff
@@ -785,5 +785,63 @@ public static IDisposable AsDisposable<T>(this T collectionIn, out T collectionO
...
+            if (sortedCollection.Count == 1)
+            {
+                return sortedCollection[0];
+            }
+            var idx = sortedCollection.BinarySearch(element);
+            if (idx >= 0) //An exact match
+            {
+                return sortedCollection[idx];
+            }
+            idx = ~idx;
```
Comment (eitanhs) [inline Bsii.Dotnet.Utils/Collections/CollectionExtensions.cs:815]: Add Comment :)

Comment (gravikovich) [reply Bsii.Dotnet.Utils/Collections/CollectionExtensions.cs]: Done

**Pull request merged**